keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,1128 @@
+is a [deprecated](https://github.com/kubernetes/enhancements/pull/3096) alpha
+feature and doesn't need to be supported anymore.
+
+A new package `k8s.io/klogr` will replace both `k8s.io/klog` and
+`k8s.io/klog/klogr`. It will be hosted under `staging` and thus get developed
+together with the code that uses it, without a stable API guarantee. It’s
+content is described in the following sections.
+
+A new staging repo is most suitable because placing the package in some of the
+other potential places would create cyclic dependencies at the module level
+(`k8s.io/component-helper` -> `k8s.io/client-go` -> `k8s.io/component-helper`),
+something that `hack/update-vendor.sh` complains about, or force consumers to
+deal with an unnecessarily large module (`k8s.io/component-base`).
+
+### Feature gate
+
+The `k8s.io/klogr` package itself cannot check Kubernetes feature gates because
+it has to be a stand-alone package with very few dependencies. Therefore it
+will have a global boolean for enabling contextual logging that programs with
+Kubernetes feature gates must set:
+
+```
+// EnableContextualLogging controls whether contextual logging is enabled.
+// By default it is enabled. When disabled, FromContext avoids looking up
+// the logger in the context and always returns the fallback logger.
+// LoggerWithValues, LoggerWithName, and NewContext become no-ops
+// and return their input logger respectively context. This may be useful
+// to avoid the additional overhead for contextual logging.
+//
+// Like SetFallbackLogger this must be called during initialization before
+// goroutines are started.
+func EnableContextualLogging(enabled bool) {
+	contextualLoggingEnabled = enabled
+}
+```
+
+The `ContextualLogging` feature gate will be defined in `k8s.io/component-base`
+and will be copied to klogr during the `InitLogs()` invocation that all
+Kubernetes commands already go through after their option parsing.
+
+`LoggerWithValues`, `LoggerWithName`, and `NewContext` are helper functions
+that wrap the corresponding functionality from `logr`:
+```
+// LoggerWithValues returns logger.WithValues(...kv) when
+// contextual logging is enabled, otherwise the logger.
+func LoggerWithValues(logger Logger, kv ...interface{}) Logger {
+	if contextualLoggingEnabled {
+		return logger.WithValues(kv...)
+	}
+	return logger
+}
+
+// LoggerWithName returns logger.WithName(name) when contextual logging is
+// enabled, otherwise the logger.
+func LoggerWithName(logger Logger, name string) Logger {
+	if contextualLoggingEnabled {
+		return logger.WithName(name)
+	}
+	return logger
+}
+
+// NewContext returns logr.NewContext(ctx, logger) when
+// contextual logging is enabled, otherwise ctx.
+func NewContext(ctx context.Context, logger Logger) context.Context {
+	if contextualLoggingEnabled {
+		return logr.NewContext(ctx, logger)
+	}
+	return ctx
+}
+```
+
+The logcheck static code analysis tool will warn about code in Kubernetes which
```

Comment:
hopefully when the gate goes away, we can undo this indirection?

Reply:
Yes. That one would be a simple regex search/replace. Added.